File: cases/mcp_package_launchers/launchers.ts
import { StdioClientTransport as Stdio } from "@modelcontextprotocol/sdk/client/stdio.js";
import { StdioClientTransport as Shadowed } from "@modelcontextprotocol/sdk/client/stdio.js";
import { StdioClientTransport as Rebound } from "@modelcontextprotocol/sdk/client/stdio.js";
import { StdioClientTransport as Fake } from "ordinary-package";

const workspace = getWorkspace();
const dynamicPackage = getPackage();

const automatic = new Stdio({
  command: "npx",
  args: ["-y", "@modelcontextprotocol/server-filesystem", workspace],
});
const pinned = new Stdio({ command: "npx", args: ["-y", "repomix@1.4.2"] });
const cached = new Stdio({
  command: "npx",
  args: ["--no-install", "@scope/local-server"],
});
const prompted = new Stdio({ command: "npx", args: ["unreviewed-server"] });
const unresolved = new Stdio({ command: "uvx", args: [dynamicPackage] });

function shadowedConstructor(Shadowed: unknown) {
  return new Shadowed({ command: "npx", args: ["-y", "shadowed-package"] });
}

Rebound = getTransport();
const rebound = new Rebound({ command: "npx", args: ["-y", "rebound-package"] });
const fake = new Fake({ command: "npx", args: ["-y", "lookalike-package"] });

const config = {
  mcpServers: {
    time: { command: "uvx", args: ["mcp-server-time"] },
    floating: { command: "npx", args: ["-y", "@x402scan/mcp@latest"] },
    pinned: { command: "npx", args: ["-y", "@scope/server@2.3.4"] },
    dynamic: { command: "uvx", args: [dynamicPackage] },
  },
};

const quoted = {
  "mcpServers": {
    "memory": { "command": "npx", "args": ["-y", "@modelcontextprotocol/server-memory"] },
  },
};

const ordinary = {
  servers: { fake: { command: "npx", args: ["-y", "ordinary-package"] } },
};

import { StdioClientTransport as SiblingPackage } from "@modelcontextprotocol/not-the-sdk";
const siblingPackage = new SiblingPackage({
  command: "npx",
  args: ["-y", "same-scope-lookalike-package"],
});

File: cases/mcp_package_launchers/launchers.py
from agents.mcp import MCPServer as Server
from agno.tools.mcp import MCPTools
from marvin.mcp import MCPServerStdio
from mcp import StdioServerParameters
from ordinary import MCPServer as FakeServer

workspace = "/workspace"
extra_args = ["--headless"]
dynamic_package = "mcp-server-fetch"

filesystem = Server(
    command="npx",
    args=["-y", "@modelcontextprotocol/server-filesystem", workspace],
)
scanner = MCPTools("npx -y @x402scan/mcp@latest")
git = MCPServerStdio(
    command="uvx",
    args=["--with", "mcp==1.29.0", "mcp-server-git"],
)

pinned_python = StdioServerParameters(
    command="uvx",
    args=["mcp-server-fetch==2026.7.10"],
)
pinned_npm = Server(command="npx", args=["-y", "repomix@1.4.2", "--mcp"])
cached_only = Server(
    command="npx",
    args=["--no-install", "@scope/local-server"],
)
prompted = Server(command="npx", args=["unreviewed-server"])
unknown = Server(command="uvx", args=[dynamic_package])
fake = FakeServer(command="npx", args=["-y", "not-an-mcp-constructor"])

config = {
    "mcpServers": {
        "time": {"command": "uvx", "args": ["mcp-server-time"]},
        "playwright": {
            "command": "npx",
            "args": ["-y", "@playwright/mcp"] + extra_args,
        },
        "pinned": {
            "command": "npx",
            "args": ["-y", "@scope/server@2.3.4"],
        },
        "dynamic": {"command": "uvx", "args": [dynamic_package]},
    }
}

ordinary_config = {
    "servers": {
        "not-mcp": {"command": "npx", "args": ["-y", "ordinary-package"]}
    }
}

from_pinned = Server(
    command="uvx",
    args=["--from", "alternate-mcp-server==1.0.0rc1", "alternate-server"],
)


def shadowed_constructor(Server):
    return Server(command="npx", args=["-y", "shadowed-package"])


def restored_module_constructor():
    return Server(command="npx", args=["-y", "restored-server@3.0.0"])
